specs/CustomDownload.md

```diff
@@ -0,0 +1,374 @@
+# Background
+<!-- TEMPLATE
+    Use this section to provide background context for the new API(s)
+    in this spec.
+
+    This section and the appendix are the only sections that likely
+    do not get copied into any official documentation, they're just an aid
+    to reading this spec.
+
+    If you're modifying an existing API, included a link here to the
+    existing page(s) or spec documentation.
+
+    For example, this section is a place to explain why you're adding this
+    API rather than modifying an existing API.
+
+    For example, this is a place to provide a brief explanation of some dependent
+    area, just explanation enough to understand this new API, rather than telling
+    the reader "go read 100 pages of background information posted at ...".
+-->
+There have been multiple requests by WebView Developers to be able to customize/control the download experience within WebView2.
+
+Some of the requests we've heard, and want to solve with this API are:
+- Developers want to be able to disable downloads for the entire environment.
+- Developers want to be able to disable downloads on a per-case basis. For example, based on the content type, or user.
+- Developers want to have access to downloaded file’s metadata.
+- Developers want to modify the download path.
+- Developers don’t want to see edge download UI if download is cancelled.
+
+In this document we describe the updated API. We'd appreciate your feedback.
+
+# Description
+
+Our proposed solution is to give developers an API to control the download experience.
+We will do this by exposing a DownloadStarting event.
+
+The DownloadStarting event will be raised when a download has begun.
```

Comment:
I'm curious. There are hosts which only allows files to be downloaded ONCE, no resume. If a developer cancels the Edge download, takes the URL and downloads it himself, would that work? Or will cancelling the download makes Edge informing the website, that the download has been aborted?

Reply:
Thanks for the question, do you mind expanding on your scenario? Is this a common scenario for you or an edge case? The way we intend for it to work is: a downloadStarting event triggers when the download begins, it would NOT cancel the download immediately. During this time, you can use the url and download metadata, to do any sort of confirmation, and block it from there. If you don't choose to block it, the download can resume as intended. Although, if you do infact block the download, expect it to work the same as cancelling a download in edge.

Reply:
I think it's an edge case as >90% of all downloads on the web allows several attempts. If I understand you right, then file will be downloaded in background anyway and the developer only have the choice to abort this download. If thats true, what happens if the file is so small, lets say 1kb, and its finishing the download before the developer has taken any action on DownloadStarting event. Is the file then present on the user's hard disk?